web-app/src/lib/db/services/credit.service.ts

```diff
@@ -1,73 +1,42 @@
-"use server";
-
-import { CreditTransactionStatus, CreditTransactionType } from "@prisma/client";
+import { CreditTransaction } from "@prisma/client";
 import { z } from "zod";
 
 import { getEnvPublicConfig } from "@/config/env.config";
 import { AgentWithFixedPricing } from "@/lib/db/extension/agent";
 import prisma from "@/lib/db/prisma";
 import { convertBaseUnitsToCredits } from "@/lib/db/utils/credit.utils";
 
-export async function getHumandReadableCreditBalance(
-  userId: string,
-): Promise<number> {
+export async function getCreditBalance(userId: string): Promise<bigint> {
   const creditBalance = await prisma.creditTransaction.aggregate({
     where: { userId },
     _sum: {
       amount: true,
     },
   });
 
-  return await convertBaseUnitsToCredits(
-    creditBalance._sum.amount ?? BigInt(0),
-  );
+  return creditBalance._sum.amount ?? BigInt(0);
 }
 
-export async function creditTransactionSpend(
+export async function createCreditTransaction(
   userId: string,
   credits: bigint,
-  includedFeeCredits: bigint,
-  note: string | null = null,
-  noteKey: string | null = null,
-) {
-  if (credits <= 0) {
-    throw new Error("Credits must be greater than 0");
-  }
-  if (includedFeeCredits < 0) {
-    throw new Error("Included fee credits must be greater than or equal to 0");
-  }
-  if (includedFeeCredits > credits) {
-    throw new Error("Included fee credits must be less than total credits");
-  }
-
-  const newCreditTransaction = await prisma.$transaction(async (tx) => {
-    const creditBalance = await tx.creditTransaction.aggregate({
-      where: { userId },
-      _sum: {
-        amount: true,
-      },
-    });
-
-    if (
-      creditBalance._sum.amount === null ||
-      creditBalance._sum.amount < credits
-    ) {
+  includedFee: bigint = BigInt(0),
+): Promise<CreditTransaction> {
+  // if credits is positive, it is a top up
+  if (credits > 0) {
+    const availableBalance = await getCreditBalance(userId);
+    if (availableBalance < credits) {
       throw new Error("Insufficient balance");
     }
+  }
 
-    return await tx.creditTransaction.create({
-      data: {
-        userId,
-        amount: -credits,
-        includedFee: includedFeeCredits,
-        type: CreditTransactionType.SPEND,
-        status: CreditTransactionStatus.PENDING,
-        note: note,
-        noteKey: noteKey,
-      },
-    });
+  return await prisma.creditTransaction.create({
```

Comment:
this should also be in a prisma transaction, as otherwise it can be a dirty read and double spends are possible

Reply:
3dcaaf6768fc89b8131cae45fa25678309a892ad